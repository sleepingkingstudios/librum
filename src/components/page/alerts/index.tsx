import * as React from 'react';
import { map } from 'lodash';

import { useAlerts } from '@alerts';
import type { Alert as IAlert } from '@alerts'
import { Alert } from '@components/alert';

const renderAlert = (alert: IAlert): JSX.Element => {
  const { uuid } = alert;

  return (
    <li key={uuid}>
      <Alert alert={alert} />
    </li>
  )
};

export const PageAlerts = (): JSX.Element => {
  const { alerts } = useAlerts();

  if (alerts.length === 0) { return null; }

  return (
    <ul className="alerts pb-3">
      { map(alerts, renderAlert) }
    </ul>
  );
};
