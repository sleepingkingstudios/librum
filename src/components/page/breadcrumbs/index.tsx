import * as React from 'react';

import { PageBreadcrumb } from './breadcrumb';
import type { Breadcrumb } from './types';

export type { Breadcrumb } from './types';

export const PageBreadcrumbs = (
  { breadcrumbs }: { breadcrumbs: Breadcrumb[] }
): JSX.Element => {
  if (breadcrumbs.length === 0) { return null; }

  return (
    <ul className="flex flex-col sm:flex-row page-breadcrumbs">
      {
        breadcrumbs.map((breadcrumb: Breadcrumb): JSX.Element => {
          const { label } = breadcrumb;

          return (
            <PageBreadcrumb key={label} breadcrumb={breadcrumb} />
          );
        })
      }
    </ul>
  )
};
