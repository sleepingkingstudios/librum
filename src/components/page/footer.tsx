import * as React from 'react';

import { FancyHr } from '@components/fancy-hr';

export const PageFooter = (): JSX.Element => (
  <footer className="mt-3 shrink-0 text-center sm:text-left text-muted">
    <FancyHr className="mb-2" />

    What lies beyond the furthest reaches of the sky?
  </footer>
);
