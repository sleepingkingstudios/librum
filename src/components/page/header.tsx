import * as React from 'react';

import { FancyHr } from '@components/fancy-hr';

interface PageHeaderProps {
  title?: string;
  subtitle?: string;
}

const renderSubtitle = ({ subtitle }: { subtitle?: string; }): JSX.Element | null => {
  if (!(typeof subtitle === 'string')) { return null; }

  return (
    <React.Fragment>
      <br className="block sm:hidden" />

      <small className="m-0 sm:ml-2 text-2xl">{ subtitle }</small>
    </React.Fragment>
  )
}

export const PageHeader = ({
  title = 'Librum',
  subtitle = 'Campaign Companion'
}: PageHeaderProps): JSX.Element => {
  return (
    <header className="font-serif leading-9 mb-3 text-5xl text-center sm:text-left text-text-strong dark:text-text-strong-dark">
      { title }

      { renderSubtitle({ subtitle }) }

      <FancyHr className="mt-1" />
    </header>
  )
}
