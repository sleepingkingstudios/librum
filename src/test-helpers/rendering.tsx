import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
