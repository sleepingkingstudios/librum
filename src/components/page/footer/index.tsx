import * as React from 'react';

import { PageBreadcrumbs } from '../breadcrumbs';
import type { Breadcrumb } from '../breadcrumbs';

const defaultClassName = 'mt-3 shrink-0 text-center sm:text-left';

const renderBreadcrumbs = ({
  breadcrumbs = [],
}: {
  breadcrumbs?: Breadcrumb[] | JSX.Element,
}): JSX.Element => {
  if (!breadcrumbs) { return null; }

  if (!('length' in breadcrumbs)) { return breadcrumbs; }

  return (<PageBreadcrumbs breadcrumbs={breadcrumbs} />);
};

export const PageFooter = ({
  breadcrumbs = [],
}: {
  breadcrumbs?: Breadcrumb[] | JSX.Element,
}): JSX.Element => (
  <footer className={defaultClassName}>
    { renderBreadcrumbs({ breadcrumbs }) }

    <hr className="hr-fancy mb-2" />

    <span className="text-muted">
      What lies beyond the furthest reaches of the sky?
    </span>
  </footer>
);
