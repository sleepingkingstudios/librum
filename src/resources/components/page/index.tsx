import * as React from 'react';

import type {
  ResourcePageOptions,
  ResourcePageProps,
} from './types';
import { responseWithStatus } from '@api';
import type { Response } from '@api';
import type { ButtonProps as Button } from '@components/button';
import { Page } from '@components/page';
import type { Breadcrumb } from '@components/page';
import {
  buildComponent,
  isElement,
} from '@utils/react-utils';
import { ResourcePageBreadcrumbs } from './breadcrumbs';
import { ResourcePageHeading } from './heading';

export type {
  ResourcePageOptions,
  ResourcePageProps,
} from './types';

const renderBreadcrumbs = ({
  action,
  breadcrumbs,
  page,
  resourceName,
  response,
  scope,
  singularName,
}: {
  action: string,
  breadcrumbs?: Breadcrumb[],
  page: ResourcePageOptions,
  resourceName: string,
  response: Response,
  scope?: string,
  singularName?: string,
}): JSX.Element => {
  if (typeof page.breadcrumbs === 'string' || isElement(page.breadcrumbs)) {
    return page.breadcrumbs;
  }

  const data = (response.data || {}) as Record<string, unknown>;
  const { status } = response;
  const pageBreadcrumbs = page.breadcrumbs as Breadcrumb[];

  return (
    <ResourcePageBreadcrumbs
      action={action}
      breadcrumbs={breadcrumbs}
      data={data}
      page={{ breadcrumbs: pageBreadcrumbs, member: page.member }}
      resourceName={resourceName}
      scope={scope}
      singularName={singularName}
      status={status}
    />
  );
};

const renderContent = ({
  page,
  ...config
}: ResourcePageProps): JSX.Element => {
  if (!('contents' in page)) { return null; }

  const {
    afterContents,
    beforeContents,
    contents,
  } = page;
  const props: Record<string, unknown> = { page, ...config };

  return (
    <>
      { buildComponent(beforeContents, props) }

      { buildComponent(contents, props) }

      { buildComponent(afterContents, props) }
    </>
  );
};

const renderHeading = ({
  action,
  buttons,
  heading,
  member,
  resourceName,
  response,
  singularName,
}: {
  action: string,
  buttons?: Button[],
  heading?: string | JSX.Element,
  member: boolean,
  resourceName: string,
  response: Response,
  singularName?: string,
}): JSX.Element => {
  if (heading && typeof heading !== 'string') { return heading; }

  const data = (response.data || {}) as Record<string, unknown>;
  const { status } = response;
  const label = heading as string;

  return (
    <ResourcePageHeading
      action={action}
      buttons={buttons}
      data={data}
      label={label}
      member={member}
      resourceName={resourceName}
      singularName={singularName}
      status={status}
    />
  );
};

export const ResourcePage = ({
  page,
  response = responseWithStatus({ status: 'uninitialized' }),
  ...config
}: ResourcePageProps): JSX.Element => {
  const {
    buttons,
    heading,
    member,
    navigation,
    subtitle,
    title,
  } = page;
  const {
    action,
    breadcrumbs,
    resourceName,
    singularName,
    scope,
  } = config;
  const renderedBreadcrumbs = renderBreadcrumbs({
    action,
    breadcrumbs,
    page,
    resourceName,
    response,
    singularName,
    scope,
  });

  return (
    <Page
      breadcrumbs={renderedBreadcrumbs}
      navigation={navigation}
      subtitle={subtitle}
      title={title}
    >
      {
        renderHeading({
          action,
          buttons,
          heading,
          member,
          resourceName,
          response,
          singularName,
        })
      }

      { renderContent({ page, response, ...config }) }
    </Page>
  );
};
