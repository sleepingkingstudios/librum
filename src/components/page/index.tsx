import * as React from 'react';

import { PageAlerts } from './alerts';
import type { Breadcrumb } from './breadcrumbs';
import { PageFooter } from './footer';
import { PageHeader } from './header';
import type { NavigationProps } from './navigation';
import { useTheme } from '@themes';
import { joinClassNames } from '@utils/react-utils';

export type {
  Breadcrumb,
} from './breadcrumbs';
export type { NavigationProps } from './navigation';

interface PageProps {
  title?: string;
  subtitle?: string;
  breadcrumbs?: Breadcrumb[] | JSX.Element;
  children: React.ReactNode,
  navigation?: NavigationProps | JSX.Element;
}

const defaultClassName =
  "flex flex-col max-w-screen-lg min-h-screen mx-auto p-3";

export const Page = ({
  title,
  subtitle,
  breadcrumbs,
  children,
  navigation,
}: PageProps): JSX.Element => {
  const joinedClassNames = joinClassNames(
    defaultClassName,
    useTheme().className,
  );

  return (
    <div className={joinedClassNames}>
      <PageHeader
        title={title}
        subtitle={subtitle}
        navigation={navigation}
      />

      <main className="grow">
        <PageAlerts />

        { children }
      </main>

      <PageFooter breadcrumbs={breadcrumbs} />
    </div>
  );
};
