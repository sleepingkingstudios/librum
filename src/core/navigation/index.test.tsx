import * as React from 'react';

import '@testing-library/jest-dom';

import { CoreNavigation } from './index';
import { Page } from '@components/page';
import { render } from '@test-helpers/rendering';

// eslint-disable-next-line @typescript-eslint/no-unsafe-return
jest.mock('@components/page', () => require('@components/page/mocks'));
jest.mock(
  '@components/page/navigation',
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  () => require('@components/page/navigation/mocks'),
);

const PageWithNavigation = () => (
  <Page navigation={<CoreNavigation />}>
    This space unintentionally left blank.
  </Page>
);

describe('<CoreNavigation />', () => {
  it('should match the snapshot', () => {
    const { asFragment } = render(
      <PageWithNavigation />,
      { router: true },
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
