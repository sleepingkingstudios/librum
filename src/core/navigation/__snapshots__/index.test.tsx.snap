// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CoreNavigation /> should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="mock-page"
  >
    <ul
      aria-label="navigation"
    >
      <li
        aria-label="navigation-item"
      >
        Home @ /
      </li>
      <li
        aria-label="navigation-item"
      >
        Publishers @ /publishers
      </li>
      <li
        aria-label="navigation-item"
      >
        Game Systems @ /game_systems
      </li>
      <li>
        Sources
        <ul>
          <li
            aria-label="navigation-item"
          >
            Books @ /sources/books
          </li>
          <li
            aria-label="navigation-item"
          >
            Websites @ /sources/websites
          </li>
        </ul>
      </li>
      <li>
        Admin
        <ul>
          <li
            aria-label="navigation-item"
          >
            Users @ /admin/users
          </li>
        </ul>
      </li>
    </ul>
    This space unintentionally left blank.
  </div>
</DocumentFragment>
`;
